subprojects/core/src/main/java/org/gradle/caching/internal/tasks/TaskOutputCacheCommandFactory.java

```diff
@@ -103,16 +133,56 @@ public BuildCacheKey getKey() {
             return new BuildCacheLoadCommand.Result<TaskOutputOriginMetadata>() {
                 @Override
                 public long getArtifactEntryCount() {
-                    return unpackResult.entries;
+                    return unpackResult.getEntries();
                 }
 
                 @Override
                 public TaskOutputOriginMetadata getMetadata() {
-                    return unpackResult.originMetadata;
+                    return unpackResult.getOriginMetadata();
                 }
             };
         }
 
+        private void updateSnapshots(ImmutableListMultimap<String, FileSnapshot> propertiesFileSnapshots) {
+            ImmutableSortedMap.Builder<String, FileCollectionSnapshot> propertySnapshotsBuilder = ImmutableSortedMap.naturalOrder();
+            for (ResolvedTaskOutputFilePropertySpec property : outputProperties) {
+                String propertyName = property.getPropertyName();
+                File outputFile = property.getOutputFile();
+                if (outputFile == null) {
+                    propertySnapshotsBuilder.put(propertyName, FileCollectionSnapshot.EMPTY);
+                    continue;
+                }
+                List<FileSnapshot> fileSnapshots = propertiesFileSnapshots.get(propertyName);
+
+                CollectingFileCollectionSnapshotBuilder builder = new CollectingFileCollectionSnapshotBuilder(UNORDERED, OutputPathNormalizationStrategy.getInstance(), stringInterner);
+                for (FileSnapshot fileSnapshot : fileSnapshots) {
+                    builder.collectFileSnapshot(fileSnapshot);
+                }
+                propertySnapshotsBuilder.put(propertyName, builder.build());
+
+                switch (property.getOutputType()) {
+                    case FILE:
+                        FileSnapshot singleSnapshot = Iterables.getOnlyElement(fileSnapshots, null);
+                        if (singleSnapshot != null) {
+                            if (singleSnapshot.getType() != FileType.RegularFile) {
+                                throw new IllegalStateException(String.format("Only a regular file should be produced by unpacking property '%s', but saw a %s", propertyName, singleSnapshot.getType()));
+                            }
+                            fileSystemMirror.putFile(singleSnapshot);
+                        } else {
+                            fileSystemMirror.putFile(new MissingFileSnapshot(outputFile.getAbsolutePath(), RelativePath.EMPTY_ROOT));
+                        }
+                        break;
+                    case DIRECTORY:
+                        Collection<FileSnapshot> descendants = Collections2.filter(fileSnapshots, EXCLUDE_ROOT_SNAPSHOTS);
+                        fileSystemMirror.putDirectory(new DirectoryTreeDetails(outputFile.getAbsolutePath(), descendants));
```

Comment:
OK. As long as there are no performance regressions that's fine with me. The regressions would be seen in our fully cached tests I guess.